### PR TITLE
Add a file to spot-check the specs of primitive floats

### PR DESCRIPTION
**Pull request by JasonGross**
- [x] Added / updated **test-suite**.
- [ ] ~Added **changelog**. (not needed, right?)~
- [x] remove dependency on #18033 (won't be merged)

### CODE REVIEW

**Comment by proux01** [comment]
LGTM in principle although I have to look at the details. Could it be expanded so as to cover the already existing tests so that we could remove them and avoid having to run and maintain both?

**Comment by JasonGross** [comment]
> Could it be expanded so as to cover the already existing tests so that we could remove them and avoid having to run and maintain both?

For many of them yes.  It's just a matter of adding the relevant constants to the list of tricky_floats or tricky_ints.  Other files seem to test things that are not about the spec, for example `Fail Check (eq_refl : zero = neg_zero).` cannot be covered.  And there are some tests covering `Z.ldexp`, which I left out in part because it's defined rather than primitive; should I extend the file to cover `Z_frexp_spec` and `Z_ldexp_spec`, even though these are proven rather than assumed?  (Also, maybe we want to leave some of the tests in there for explanatory value?)

**Review by proux01 — COMMENTED** [review]
Thanks for the pull request and sorry for the late review (I definitely don't have your Ltac experience and all that Ltac code appeared a bit daunting at first glance ;-) ). There are a few points that could be improved but I really think this deserves polishing and merging.

**Comment by proux01** [comment]
> > Could it be expanded so as to cover the already existing tests so that we could remove them and avoid having to run and maintain both?
> 
> For many of them yes. It's just a matter of adding the relevant constants to the list of tricky_floats or tricky_ints. Other files seem to test things that are not about the spec, for example `Fail Check (eq_refl : zero = neg_zero).` cannot be covered. And there are some tests covering `Z.ldexp`, which I left out in part because it's defined rather than primitive;

@JasonGross I agree

> should I extend the file to cover `Z_frexp_spec` and `Z_ldexp_spec`, even though these are proven rather than assumed?

No, this doesn't seem very useful indeed

> (Also, maybe we want to leave some of the tests in there for explanatory value?)

You're right, your code only tests consistency of spec with all evaluation mechanisms, it is still good to check that the spec does what we expect on a few given values.

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v:321]: That is a good idea, I will go and implement this.

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v:93]: https://github.com/coq/coq/blob/6798c64f4fbf5cbf0a68062f627286a2f9a8beb2/theories/Init/Tactics.v#L324-L330

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v:94]: Indeed.  I guess the unfortunate thing here is that the startup time of the native-compiler is so slow that we can't check all specs individually and instead want to reduce them all at once, so we can't factor the code between `hnf` and `native_compute`.  Maybe I can have the `hnf` variant iterate down the list in Ltac and run `hnf` on each one individually?

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v:321]: Unfortunately even the LHS of specs includes a decent amount of reasoning about spec_float (in `Prim2SF`), and `cbn` and `simpl` are slow enough that I don't think we should test the entire set of examples on them.  But I can test a reduced set of examples (and, indeed, I've uncovered a bug in `simpl` in doing so)

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v]: Integrated, thanks!

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v]: Integrated, thanks!

**Comment by JasonGross** [comment]
@proux01 I've rewritten all the machinery in Ltac2 for (hopefully) ease of maintainability and understandability.  (@ppedrot may tell me that I'm doing arcane unsafe things with Ltac2 and should be writing things differently though)
I've also added a bunch of extensive documentation, added support for `simpl`, `cbn`, and `hnf`, and caught a couple bugs in `simpl` / `hnf`, which I will report momentarily.

> > > Could it be expanded so as to cover the already existing tests so that we could remove them and avoid having to run and maintain both?
> > 
> > 
> > For many of them yes. It's just a matter of adding the relevant constants to the list of tricky_floats or tricky_ints. Other files seem to test things that are not about the spec, for example `Fail Check (eq_refl : zero = neg_zero).` cannot be covered. And there are some tests covering `Z.ldexp`, which I left out in part because it's defined rather than primitive;
> 
> @JasonGross I agree
> 
> > (Also, maybe we want to leave some of the tests in there for explanatory value?)
> 
> You're right, your code only tests consistency of spec with all evaluation mechanisms, it is still good to check that the spec does what we expect on a few given values.

Having looked through the tests again, it's not so easy for me to tell which tests are testing the specs and which are testing the implementations.  If you'd like, I can update `specs.v` so that it covers every single test of a spec in the other files, by adding all the constants to the relevant lists.  Should I do that?

Regardless, I'd like to leave it to someone else to decide which tests are "of specs" and which are "of implementations" and remove the ones that are just testing that the implementations match the specs.

**Comment by JasonGross** [comment]
> I can update `specs.v` so that it covers every single test of a spec in the other files, by adding all the constants to the relevant lists. Should I do that?

I've done it.  The combinatorial explosion makes things a bit slower (by a minute or three), so I'm not sure if it's a good change.

**Comment by JasonGross** [comment]
@proux01 CI passes and this is ready for re-review

File: test-suite/primitive/float/specs.v
Comment (proux01) [reply test-suite/primitive/float/specs.v]: Well, for most axioms (except maybe the three first ones `Prim2SF_valid`, `SF2Prim_Prim2SF` and `Prim2SF_SF2Prim`), the computations of `Prim2SF` is not really what we want to test. So maybe for axioms like
```Coq
Axiom opp_spec : forall x, Prim2SF (-x)%float = SFopp (Prim2SF x)
```
it would be easier to precompute `SF2Prim (SFopp (Prim2SF x))` with `vm_compute` and only check that this matches `(-x)%float` on each evaluation mechanism. To be on the safe side, one could also check with `vm_compute` that `valid_binary (SFopp (Prim2SF x))` holds.

**Comment by proux01** [comment]
> @proux01 CI passes and this is ready for re-review

Sorry for the late answer (vacation period here in France), I hope to do a careful review by the end of the month.

> Regardless, I'd like to leave it to someone else to decide which tests are "of specs" and which are "of implementations" and remove the ones that are just testing that the implementations match the specs.

Sure, I will do that.

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v]: Maybe a better way to check simpl and cbn is just to make a list of the primitive floating point operations, and check for a constr_eq match between the various reduction mechanisms, and only check against the specs for VM, cbv, lazy, native.

Comment (proux01) [reply test-suite/primitive/float/specs.v]: If that's easier to implement sure. That sounds equivalent to me but I must admit I did don dived into the implementation details.

**Comment by JasonGross** [comment]
I've pushed an update that runs cbn and simpl on just the primitive operations, on all the edge cases.  Unfortunately hnf is still too slow, because iterating over the list of things to reduce in Ltac (to invoke hnf on each one) is pretty slow.  This is strictly a code addition, because the previous code still needs to be there for `hnf`, and because the previous code did reveal bugs in `simpl` that the new code does not reveal (related to how `simpl` treats results of floating point primitives when reducing larger expressions).

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [reply test-suite/primitive/float/specs.v]: Mostly it's about not having to rely on the structure of specs being `Prim2SF ... = ...`.  Since really what we're checking is that the primitive float operations compute the same across reduction mechanisms, we can just check that directly, and then separately check that this matches the spec using only the fastest of reduction mechanisms.

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [inline test-suite/primitive/float/specs.v:284]: @proux01 While this way is simpler, I worry that it's more likely to get out of sync with the axioms in the standard library, and the extra complexity in testing might be worth paying to ensure that any changes in the spec in the stdlib get propagated here.  What do you think?

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [inline test-suite/primitive/float/specs.v:454]: Should this be timed?

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [inline test-suite/primitive/float/specs.v:457]: This should be timed too?

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [inline test-suite/primitive/float/specs.v:460]: This should also be timed

File: test-suite/primitive/float/specs.v
Comment (JasonGross) [inline test-suite/primitive/float/specs.v:463]: This should also be timed

**Comment by JasonGross** [comment]
Thanks for the simplifications!

> You're right that it no longer checks `simpl` as extensively but, strictly speaking, that's not a test of the floating point specification. Maybe we should put a few of those tests in yet another file?

I was thinking that we could leave them in here as they are "spec-derived", but having a couple of dedicated tests in another file (to be added in another PR?) also seems fine.

**Comment by JasonGross** [comment]
But in any case, I am happy with your edits, and would love to see this merged when ready.

File: test-suite/primitive/float/ldexp.v
Comment (JasonGross) [inline test-suite/primitive/float/ldexp.v:6]:
```suggestion
```

File: test-suite/primitive/float/specs.v
Comment (proux01) [reply test-suite/primitive/float/specs.v]: No strong opinion. I'm not really afraid of axioms getting out of sync as those are not expected to change, plus there is already `Prim2SF_SF2Prim` that requires a similar specific treatment. Anyway, reverting.

File: test-suite/primitive/float/specs.v
Comment (proux01) [reply test-suite/primitive/float/specs.v]: It's so fast that I did not deemed it useful but why not.

File: test-suite/primitive/float/ldexp.v
Comment (proux01) [reply test-suite/primitive/float/ldexp.v]: Indeed, thanks!

**Comment by proux01** [comment]
> > You're right that it no longer checks `simpl` as extensively but, strictly speaking, that's not a test of the floating point specification. Maybe we should put a few of those tests in yet another file?
> 
> I was thinking that we could leave them in here as they are "spec-derived", but having a couple of dedicated tests in another file (to be added in another PR?) also seems fine.

It seems unlikely this kind of bugs of simpl are triggered by a very specific combination of values and operators, as tested in the new `specs.v`, so IMHO a few tests in another file should be enough (and much cheaper).

**Review by proux01 — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Commit history looks like it could use some squashing

**Comment by proux01** [comment]
> Commit history looks like it could use some squashing

Indeed, I'll do some squashing before merging.

**Comment by JasonGross** [comment]
> Unless I am mistaken, there are no negative tests. In particular, I am unsure about what happens if a bug were to be introduced.

This behavior is controlled by `Ltac2 report_result`.  Previously, we were testing the behavior of `simpl` on more complex operations (like `Prim2SF ...`), and this uncovered https://github.com/coq/coq/issues/17932.  I'm this case (a reduction mechanism fails to fully reduce), `report_result` uses `Control.throw` to report an error like
```
fprintf "%s failed to fully reduce, leaving over %t (expected something equivalent to: %t), in %t (%t)" red lhs rhs spec specTy
```
If the computation gives the wrong result (something that fails to unify with the expected result, rather than something that unifies but is not constr_eq), the error message instead looks like:
```
fprintf "%s failed!%sGot: %t%sExpected something equivalent to: %t%s(both sides %s unify)%sIn %t (%t)" red (lf ()) lhs (lf ()) rhs (lf ()) descr (lf ()) spec specTy
```

**Comment by proux01** [comment]
Maybe the easiest way to add some negative test would be to `Fail Check` an erroneous spec like `forall x, abs x = - x` for instance.

**Comment by proux01** [comment]
@silene I added a test of the testing machinery in the last commit, does that correspond to what you were expecting?

@JasonGross I'm a bit surprised by the precise error message though (I wasn't expecting the "failed to fully reduce" as the result is indeed fully reduced)

**Comment by erikmd** [comment]
@proux01 

> I'd like to merge by the end of the week if there is no further comment.
>
> @erikmd and @silene you may want to hear about this

Thanks! I'll try to have a look before the end of the week

**Comment by JasonGross** [comment]
> @JasonGross I'm a bit surprised by the precise error message though (I wasn't expecting the "failed to fully reduce" as the result is indeed fully reduced)

I guess we need to replace `Ltac2 Eval` with `Check ltac2:(...)`, because `Ltac2 Eval` is broken (https://github.com/coq/coq/issues/18021)

**Comment by proux01** [comment]
@JasonGross this doesn't seem to work (or is way too slow) so I just commented out the more elaborate message for now.

**Comment by JasonGross** [comment]
> @JasonGross this doesn't seem to work (or is way too slow) so I just commented out the more elaborate message for now.

Seems very weird for it to be meaningfully slower.  Even when there's just one spec, it hangs?  What about doing `Goal False. Ltac2 Eval ... Abort.`?  I do think there's useful information to be had in the more elaborate message.

**Comment by proux01** [comment]
@JasonGross this seems to work but the `Goal False` itself is way too slow after all the `Let` imposed by `native_compute`.

**Comment by JasonGross** [comment]
Oh, ew, I see.  Okay, commenting it out for now it is.

**Comment by erikmd** [comment]
Hi! thanks @JasonGross for working on this 👍
@proux01 could you wait just a little bit more before merging? (I'd like to take a look but I won't be able to do it before Thursday…)

**Comment by proux01** [comment]
@erikmd this has just been rebased on top of https://github.com/coq/coq/pull/18033 and is thus waiting for it to be merged, so I guess you have plenty of time now.

**Comment by JasonGross** [comment]
Alternatively, if we want to merge this sooner, I can move the refactoring to use https://github.com/coq/coq/pull/18033 to another PR?

**Comment by proux01** [comment]
Well, we are not in a hurry and there are not much risks of conflicts so, except if it takes forever for #18033 to be merged, le's wait.

**Comment by proux01** [comment]
@JasonGross since it seems more and more unlikely https://github.com/coq/coq/pull/18033 is going to land soon, what about merging this as it is?

**Comment by SkySkimmer** [comment]
#18033 isn't going to land for sure, I'm not going to continue it

**Comment by proux01** [comment]
Thanks for the information. Then let's merge, this PR has been open for way too long.

@coqbot merge now

**Pull request merged**
